src/main/scala/com/github/symcal/Expr.scala

```diff
@@ -261,28 +267,38 @@ final case class Sum(es: Expr*) extends Expr {
   }
 
   override def simplify: Expr = {
-    val (const, nonconst) = es.map(_.simplify).partition(_.isConst)
-    // mergedConstants is always non-empty but can be Const(0)
-    val mergedConstants = const.foldLeft[Expr](Const(0))((x, y) ⇒ Const(x.toInt + y.toInt))
-    val mergedExprs = mergedConstants match {
-      case Const(0) ⇒ nonconst
-      case _ ⇒ nonconst :+ mergedConstants
+    val (constants, nonconstants) = es.map(_.simplify).partition(_.isConst)
+    val nonconstantsFlattened: Seq[Expr] = nonconstants.flatMap {
+      case Sum(es@_*) ⇒ es
+      case x => Seq(x) // not a sum
+    }
+    // mergedConstants is the product of all constants in the list; also could be 0 or 1.
+    val mergedConstants: Int = constants
+      .collect { case x@Const(_) ⇒ x } // This converts into Seq[Const]. We know that we are not losing any values here.
+      .map(_.value) // Here we can just get _.value of the `Const`.
+      .reduceOption(_ + _) // This may yield `None` if sequence is empty.
+      .getOrElse(0) // An empty list of `constants` will also produce 0 here.
+    val mergedExprs: Seq[Expr] = mergedConstants match {
+      case 0 ⇒ nonconstantsFlattened
+      case _ ⇒ nonconstantsFlattened :+ Const(mergedConstants)
     }
     // There are three cases now: empty sequence, one expr, and more than one expr.
     mergedExprs.headOption match {
       case Some(e) ⇒
         if (mergedExprs.length == 1) {
-          // In this case, the simplified result is not a `Sum`.
+          // In this case, the simplified result is not a `Product`.
```

Comment:
what do you mean by `// In this case, the simplified result is not a `Product`.`?